feat: add `namespace` option to generate `.d.ts` files

aviemet opened the pull request:
Hello, I'd like to propose that an option for changing the file extension be made available. I took the liberty of making a PR with my suggestion. Would love to know what you think.

This adds a config option to set whether the output file should have a ".ts" or ".d.ts" extension.

Thank you

aviemet commented:
So, first off, sorry for the initial PR comment. I meant to only merge to my fork while I was playing around, but accidentally did a PR with no subject or commit message, and then I kinda panicked and edited one in real quick.

What I wanted to propose was an option to generate declaration files rather than ts files. In my project, I'm already using types defined in a global namespace, so having the namespace generated would save a ton of work adding imports to each file. I initially tried just importing the types into a declaration file and exporting them under a namespace declaration, but after a lot of trial and error, my conclusion was that there is an issue with the module system which hinders re-exporting types which themselves were default exports from a file. Sadly, I couldn't find any documentation about this, but changing the default exports to named exports made things work.

Anyway, I tried to make my proposed changes flexible by adding a "namespace" option to the initializer, which if not set generates ts files as normal, but if given a string value, generates d.ts files instead. It would declare a global namespace containing all of the generated types, making them globally available without an import statement. This did require more changes than I anticipated, so I hope that I'm not proposing something which is over-fit to my specific needs.

I'd love to know if this is a feature you feel is worth including.

Thanks again for the great project!

ElMassimo commented:
Hi Avram!

I like the addition of `namespace`, I think for certain types of projects it's reasonable to have global types.

For large projects, it's probably better to import types manually, though in those cases it might be viable to trigger generate separately for "areas" of the app, and use a different namespace for each.

---

I'd rather avoid the `indentation` setting, would you remove it from this PR?

I'm ok switching the default to `tabs`, but would prefer a separate PR for that change (maybe any users watching the issue can vote).

aviemet commented:
Sure thing, I removed the indentation option. I had added it in part because I happen to enforce tabs for my frontend files and eslint made a bunch of squigglies when peeking at the generated files (which is very much a non-issue).

Also, a quick note about the snapshot tests. I had to change how they are generated because on my Linux build, the "::" character sequence wasn't allowed in a filename. So this also includes a change to the snapshot generation which converts those to "__". I didn't include the updated snapshots because I didn't want to muddy the PR with a bunch of changes, but can do so if you'd like.

aviemet commented:
Hello, checking in to see if there's something else I do to make this PR work for you. Let me know, happy to do some refactoring or add some tests.

ElMassimo commented:
Hi Avram! Life's been busy around here (moved to a new place).

I intend to merge and release this PR, but need to find some time to test it myself. Thanks!

ElMassimo commented on spec/types_from_serializers/__snapshots__/interfaces_ComposerSerializer.snap:
```diff
@@ -2,10 +2,9 @@
 // DO NOT MODIFY: This file was automatically generated by TypesFromSerializers.
 
-export default interface Composer {
+export interface Composer {
```
The "standard" mode should not be affected by this PR, as it's a breaking change, and I'd like to ship namespace support in a minor release.

Let's keep the `default` export when `namespace` is `nil`.

ElMassimo commented on types_from_serializers/lib/types_from_serializers/generator.rb:
```diff
@@ -412,17 +421,34 @@ def serializers_index_content(serializers)
+      config.namespace ? declaration_interface_definition(interface) : standard_interface_defenition(interface)
+    end
+
+    def standard_interface_defenition(interface)
```
Small typo here:

```suggestion
    def standard_interface_definition(interface)
```

aviemet commented:
Funny, I actually just moved as well, and then immediately went on holiday. I believe this last commit addresses your points above, avoiding breaking changes makes a lot of sense for a minor release.

aviemet commented:
Is there anything else I can do to get this ready? I've been using it in my project and haven't had any issues, I'm referencing the branch directly in my Gemfile. I'd really like to get this rolled in to the gem because CircleCI is trying to run the tests in this project during my build process and failing.

ElMassimo commented:
Hi! [This comment is still not addressed](https://github.com/ElMassimo/types_from_serializers/pull/9/files/41e291910d39c73924a8771c45354b5314413d29#r1180502697), which I think makes this change backwards-incompatible as is.

We'd want no changes to "normal" snapshots, plus all the new ones for "namespaced" tests.

aviemet commented:
I've never used Github comments before, so maybe I resolved them incorrectly, but I believe my last commit addressed all of the concerns. Exports are changed back to default, typo is corrected. I reverted the tests to assert default exports as well, and they are passing.

ElMassimo commented:
If you check [this import](https://github.com/ElMassimo/types_from_serializers/pull/9/files#diff-19522f3092a3a24618aff176f4edc302bb3747fdcb05c9a40b882015bdc29726R4) for example, it will fail under `tsc`, because there is no `Model` export in that file.

> Exports are changed back to default

Correct, but the imports are still using named imports.

ElMassimo commented:
Thanks for your contribution Avram!

ElMassimo commented:
Released in `types_from_serializers@2.1.0` 🎉

aviemet commented:
Hey, thank you for letting me contribute to this and for your patience while I figured things out. I'm not a professional developer, this is more of a hobby for me, so it was really fun to move through the steps of proposing and contributing a new feature.
